SO-6331: Support Java 21

cmark opened the pull request:
Jira: https://snowowl.atlassian.net/browse/SO-6331

cmark committed:
ci: update github flow to use JDK 21

cmark committed:
feat(tp): update tp to include Java 21 in certain target envs

cmark committed:
build: set compiler target to java release 21

cmark committed:
fix(core): potential resource leak when SSL cert allocation fails

cmark committed:
fix(core): log warn when self-signed certificate gen fails

cmark committed:
feat(launch): ensure that requiredJavaVersion is set to 21

nagyo reviewed (changes requested):
We need changes in the Jenkinsfile and some potential configuration changes in Jenkins itself in order to let this change go through.

cmark commented:
> We need changes in the Jenkinsfile and some potential configuration changes in Jenkins itself in order to let this change go through.

Yup, that's on the todo list as well, but at the moment we currently have one blocker issue, so I've put this into draft mode. Please wait with any reviews until I mark it ready to review. Thx!

cmark committed:
Merge branch '9.x' into upgrade/java-21

cmark committed:
chore(dev): migrate core.dependencies feature to newer format

cmark committed:
feat(deps): add bouncycastle 1.77 jars to target platform

cmark committed:
feat(build): add bouncycastle 1.77 jars to core.deps feature

cmark committed:
feat(core)!: do not rely on sun.security.x509 packages...

...when generating self-signed certificate

Use bouncycastle instead which is picked up by netty during startup.
Remove all add-opens from product definitions and startup scripts.

cmark committed:
chore(dev): update launch configurations

Bump to Java-21 execution environment.
Remove sun.security.x509 add-opens directives.

cmark committed:
feat(build): update exec env of bundles to Java-21

cmark committed:
test(eventbus): ensure bouncycastle modules are loaded during tests

cmark committed:
feat(deps): upgrade Groovy to 3.0.24

cmark committed:
chore(mapdb): bump version of mapdb wrapper to 3.1.1.b2i

cmark committed:
chore(launch): remove org.apache.log4j in favor of log4j.over.slf4j

cmark committed:
fix(build): update mapdb version ref in snomed.core.feature

cmark committed:
ci(jenkins): use jdk21 agents

apeteri reviewed (approved):
Projects compile cleanly, which is promising 🤞 

Please take a second peek at the plug-in selection of each launch configuration and check:

- whether they need short-form Bouncy Castle bundles (`bcpg`, `bcprov`, `bcpkix`, `bcutil`)
- whether fully qualified BC bundles are still present (`org.bouncycastle.bcpg`, `org.bouncycastle.bcprov`)

For the second set I only found these occurrences: https://github.com/b2ihealthcare/snow-owl/blob/6a3b252f602dccb380550d40cd91a9deae803127/fhir/com.b2international.snowowl.fhir.rest/.launch/fhir-api-server.launch#L95-L96

The following launch configurations show a dialog indicating missing dependencies instead of opening a Console view and starting:

- `fhir-api-server.launch`
- `snomed-e2e-tests (cis).launch`
- `snomed-e2e-tests (es8).launch` &rarr; this works after double-checking
- `snow-owl-cis.launch`
- `snow-owl-oss.launch`

👍 otherwise.

apeteri commented:
Apologies, I'll check the launch configs again tomorrow! My IDE was not up to date.

apeteri commented:
After making sure the IDE has a JRE 21 installed only 4 out of 5 configs remain.

cmark committed:
chore(dev): resolve remaining launch configuration issues

apeteri reviewed (approved):
✅

nagyo reviewed (approved):
LGTM!
